packages/flutter/lib/src/rendering/flow.dart

```diff
@@ -0,0 +1,345 @@
+
+  @override
+  double getMaxIntrinsicHeight(BoxConstraints constraints) {
+    return _getSize(constraints).height;
+  }
+
+  @override
+  void performLayout() {
+    size = _getSize(constraints);
+    int i = 0;
+    RenderBox child = firstChild;
+    while (child != null) {
+      BoxConstraints innerConstraints = _delegate.getConstraintsForChild(i, constraints);
+      child.layout(innerConstraints, parentUsesSize: true);
+      final _FlowParentData childParentData = child.parentData;
+      childParentData.offset = Offset.zero;
+      child = childParentData.nextSibling;
+      i += 1;
+    }
+  }
+
+  final List<int> _lastPaintOrder = <int>[];
+  PaintingContext _paintingContext;
+  Offset _paintingOffset;
+  List<RenderBox> _randomAccessChildren;
+
+  @override
+  Size getChildSize(int i) {
+    if (i < 0 || i >= _randomAccessChildren.length)
+      return null;
+    return _randomAccessChildren[i].size;
+  }
+
+  @override
+  void paintChild(int i, Matrix4 transform) {
+    transform ??= new Matrix4.identity();
+    _lastPaintOrder.add(i);
+    RenderBox child = _randomAccessChildren[i];
+    final _FlowParentData childParentData = child.parentData;
+    assert(() {
+      if (childParentData._transform != null) {
+        throw new FlutterError(
+          'Paint each child at most once.\n'
+          'The flow delegate of type ${_delegate.runtimeType} attempted to '
+          'paint child $i multiple times, which is not permitted.'
+        );
+      }
+      return true;
+    });
+    childParentData._transform = transform;
+    _paintingContext.pushTransform(needsCompositing, _paintingOffset, transform, (PaintingContext context, Offset offset) {
+      context.paintChild(child, offset);
+    });
+  }
+
+  void _paintWithDelegate(PaintingContext context, Offset offset) {
+    _lastPaintOrder.clear();
+    _paintingContext = context;
+    _paintingOffset = offset;
+    _randomAccessChildren = getChildrenAsList();
+    for (RenderBox child in _randomAccessChildren) {
+      final _FlowParentData childParentData = child.parentData;
+      childParentData._transform = null;
+    }
+    try {
+      _delegate.paintChildren(this);
+    } finally {
+      _paintingContext = null;
+      _paintingOffset = null;
+      _randomAccessChildren = null;
+    }
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    context.pushClipRect(needsCompositing, offset, Point.origin & size, _paintWithDelegate);
+  }
+
+  @override
+  bool hitTestChildren(HitTestResult result, { Point position }) {
+    final List<RenderBox> children = getChildrenAsList();
+    for (int i = _lastPaintOrder.length - 1; i >= 0; --i) {
+      final int childIndex = _lastPaintOrder[i];
+      if (childIndex >= children.length)
+        continue;
+      final RenderBox child = children[childIndex];
+      final _FlowParentData childParentData = child.parentData;
+      final Matrix4 transform = childParentData._transform;
+      if (transform == null)
```

Comment:
If you're not using the widgets library, the tree can be mutated arbitrarily between when it paints and when we later hit test it.